src/keyboard.rs

```diff
@@ -1664,3 +1567,209 @@ pub enum KeyLocation {
+        /// The "shift" key.
+        const SHIFT = 0b100;
+        // const LSHIFT = 0b010 << 0;
+        // const RSHIFT = 0b001 << 0;
+        /// The "control" key.
+        const CONTROL = 0b100 << 3;
+        // const LCTRL = 0b010 << 3;
+        // const RCTRL = 0b001 << 3;
+        /// The "alt" key.
+        const ALT = 0b100 << 6;
+        // const LALT = 0b010 << 6;
+        // const RALT = 0b001 << 6;
+        /// This is the "windows" key on PC and "command" key on Mac.
+        const SUPER = 0b100 << 9;
+        // const LSUPER  = 0b010 << 9;
+        // const RSUPER  = 0b001 << 9;
+    }
+}
+
+impl ModifiersState {
+    /// Returns `true` if the shift key is pressed.
+    pub fn shift_key(&self) -> bool {
+        self.intersects(Self::SHIFT)
+    }
+    /// Returns `true` if the control key is pressed.
+    pub fn control_key(&self) -> bool {
+        self.intersects(Self::CONTROL)
+    }
+    /// Returns `true` if the alt key is pressed.
+    pub fn alt_key(&self) -> bool {
+        self.intersects(Self::ALT)
+    }
+    /// Returns `true` if the super key is pressed.
+    pub fn super_key(&self) -> bool {
+        self.intersects(Self::SUPER)
+    }
+}
+
+#[cfg(feature = "serde")]
+mod modifiers_serde {
+    use super::ModifiersState;
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    #[derive(Default, Serialize, Deserialize)]
+    #[serde(default)]
+    #[serde(rename = "ModifiersState")]
+    pub struct ModifiersStateSerialize {
+        pub shift_key: bool,
+        pub control_key: bool,
+        pub alt_key: bool,
+        pub super_key: bool,
+    }
+
+    impl Serialize for ModifiersState {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let s = ModifiersStateSerialize {
+                shift_key: self.shift_key(),
+                control_key: self.control_key(),
+                alt_key: self.alt_key(),
+                super_key: self.super_key(),
+            };
+            s.serialize(serializer)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for ModifiersState {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let ModifiersStateSerialize {
+                shift_key,
+                control_key,
+                alt_key,
+                super_key,
+            } = ModifiersStateSerialize::deserialize(deserializer)?;
+            let mut m = ModifiersState::empty();
+            m.set(ModifiersState::SHIFT, shift_key);
+            m.set(ModifiersState::CONTROL, control_key);
+            m.set(ModifiersState::ALT, alt_key);
+            m.set(ModifiersState::SUPER, super_key);
+            Ok(m)
+        }
+    }
+}
+
+/// Extra information on the current [`ModifiersState`].
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ModifiersMetadata {
```

Comment:
API-wise I'd prefer:
```rust
#[non_exhaustive]
pub struct Modifiers {
    pub state: ModifiersState,
    pub(crate) pressed_mods: ModifiersKeys,
}

// Along with:
// WindowEvent::ModifiersChanged(Modifiers)
```

And in the future, even getting rid of the `pub` on `state`, in favour of explicit methods on `Modifiers`. But just a preference, the current API works fine.

Reply:
Nah, I had something like that in fist revision and it's just bad. You simply want to write stuff like `Modifiers::SUPER | Modifiers::SHIFT` and you want to have all the bitflags semantics.

You also want to have `eq` based on just `state` not the `Modifiers` in particular, given that pressed_mods/metadata doesn't provide any end value to the users.

So, the main issue is that you really want all of that `bitflags` semantics, the matadata is just optional field to check on some platforms that you have something pressed. On most it's not possible to tell which key resulted in modifiers change.

Reply:
Though, maybe I got you wrong and you simply want to have a distinct structure for the event.

Not sure if it'll be any better though, I think it'll be around the same, and `#[non_exhaustive]` is redundant due to presence of metadata.

Reply:
I just wanted a distinct struct for the event, but again, mostly preference

Reply:
Updated to that.